Fixed day shifts when marking in Welcome Modal or Tab Home

imblowfish opened the pull request:
Closes #143

The problem occured if the user's timezone differed from the UTC timezone by 1 day. In this case, ISO string was passed as parameter to IonDatetime's `isDateEnabled` or `highlightedDates`. And inside these functions the date was passed to `Date` constructor. The `Date` constructor in turn, translated this date into local time, which could case a shift.

As example. Now is October 28 01:15 in UTC. Let's imagine that user has `America/Los_Angeles` timezone.  IonDatetime passes ISO string `2023-10-28` as parameter for `isDateEnabled` callback. This string is passed as argument in the `Date` constructor and `Date` returns `Fri Oct 27 2023 17:00:00 GMT-0700 (Pacific Daylight Time)` to us. But initially the date is already described the current day from IonDatetime and we don't need to re-convert the time to the current timezone.

To get rid of this convertion I changed `Date` constructor to `parseISO` function from `date-fns`.

Also I changed this behavior in `CalculationsLogic.ts`. I tried to implement unit tests, but didn't found any possibility to change timezone in a specific tests wich Jest like we can do that in the Chrome Devtools with `Sensors`, so I added a new tests in the test plan.

IraSoro commented on src/pages/TabHome.tsx:
```diff
@@ -199,12 +200,12 @@ const TabHome = (props: HomeProps) => {
                 isDateEnabled={(date: string) => {
```
I suggest parsing here so that in ```getActiveDates``` there are only calculations.
And change  ```date: string``` argument to ```isoDateString```.

imblowfish replied:
Agree, fixed

IraSoro commented on src/state/CalculationLogics.ts:
```diff
@@ -301,8 +307,8 @@ export function getLastPeriodDays(cycles: Cycle[]) {
-export function getActiveDates(dateString: string, cycles: Cycle[]) {
-  const date = startOfDay(new Date(dateString));
+export function getActiveDates(isoDateString: string, cycles: Cycle[]) {
+  const date = startOfDay(parseISO(isoDateString));
```
Therefore, remove the ```parseISO``` here

imblowfish replied:
Fixed

IraSoro commented:
There are more sensory tests, so I will replace them into issues in the new milestone “Tests”.

IraSoro commented:
I found some bugs:
1. Array of cycles is empty. Welcome modal is open. I mark period days:
![image](https://github.com/IraSoro/peri/assets/52165881/467c5520-fa1a-4246-9ec5-71da5bec5239)

2. I click the ```continue``` button. Home tab is open. Calendar looks like this:
![image](https://github.com/IraSoro/peri/assets/52165881/2816c1cb-2ecc-4c7d-8671-14d38e4ab692)

3. But if we want to see future days of period, it will not show one day (there should be 5 marked and not 4)
![image](https://github.com/IraSoro/peri/assets/52165881/274a66f4-10ed-4bcd-b6b8-06525e0c107e)

IraSoro commented:
2. This bug follows from the first one

1. Open ```Welcome modal```. Mark this days:
![image](https://github.com/IraSoro/peri/assets/52165881/76a2fd14-9846-4c7e-bbeb-1dd36d3caa82)

2.  Here you can see that your period is today, but today is not marked
![image](https://github.com/IraSoro/peri/assets/52165881/33fa74e0-f97a-4ec8-bed5-09146a6cccab)

From these two bugs it turns out that the app doesn`t count the future first day of period

IraSoro commented:
But if the array is full (the fuction ```_todayPeriod(6).catch((err) => console.error(err));``` is uncommented), it calculates everything correctly
![image](https://github.com/IraSoro/peri/assets/52165881/a0d948f5-715c-4f25-bfc0-456a2cd04bb3)

IraSoro commented:
You can also delete the `TEST_PLAN.md` file. I created a new [milestone](https://github.com/IraSoro/peri/milestone/13) and moved all the described tests there.

imblowfish commented:
> I found some bugs: 1.
> 
> 1. Array of cycles is empty. Welcome modal is open. I mark period days:
>    ![image](https://user-images.githubusercontent.com/52165881/278882380-467c5520-fa1a-4246-9ec5-71da5bec5239.png)
> 2. I click the `continue` button. Home tab is open. Calendar looks like this:
>    ![image](https://user-images.githubusercontent.com/52165881/278882505-2816c1cb-2ecc-4c7d-8671-14d38e4ab692.png)
> 3. But if we want to see future days of period, it will not show one day (there should be 5 marked and not 4)
>    ![image](https://user-images.githubusercontent.com/52165881/278882590-274a66f4-10ed-4bcd-b6b8-06525e0c107e.png)

Fixed

**Tested with San Francisco**
![image](https://github.com/IraSoro/peri/assets/28269602/da3f5111-a9f0-47ca-9f97-8f61c226e24e)
![image](https://github.com/IraSoro/peri/assets/28269602/31957d9c-430a-4528-a834-75ddc12d6bac)
![image](https://github.com/IraSoro/peri/assets/28269602/bc82bb43-3f50-481d-8fcf-04a88a3a66df)
![image](https://github.com/IraSoro/peri/assets/28269602/0d8d2ae2-96d7-4a1d-88fe-441c20932067)

**Tested with Tokyo**
![image](https://github.com/IraSoro/peri/assets/28269602/838a2f64-6c8f-4256-bf48-6d51cd1e0192)
![image](https://github.com/IraSoro/peri/assets/28269602/14e1bd14-b04c-430b-8630-0ac7734ba725)
![image](https://github.com/IraSoro/peri/assets/28269602/6ac5fd1f-1932-4b96-9ccd-fd70587d6197)
![image](https://github.com/IraSoro/peri/assets/28269602/4d37b656-99cf-4729-8d9e-e51f2e459134)

imblowfish commented:
I fixed all problems described above and added related tests
Also, in this PR I changed cycles storage approach. Previously we stored cycle start date as "yyyy-MM-dd", but not we'll store it as local date instead

IraSoro commented:
I found a bug.
Steps:
1. Set San Francisco location
2. Set empty array
3. Open `Welcome modal` and mark these days:
![image](https://github.com/IraSoro/peri/assets/52165881/1f73a987-5f82-4c9b-9fd9-43fb5e39ace0)

4. Click `continue` and move to the `tab Home`:
![image](https://github.com/IraSoro/peri/assets/52165881/ba6528df-1df9-4242-9a49-aaa22c6cdcf2)

5. Click `mark button`
![image](https://github.com/IraSoro/peri/assets/52165881/2c019c52-74f5-4c21-9338-c0a32903febc)

There was a shift back 1 day. This only happens in the San Francisco location. 

And the array can be made not empty. You can, for example, uncomment the `_randomLutealPhase(6)` function (or `_randomDelayOfCycle(6) or _todayPeriod(6));` function):
![image](https://github.com/IraSoro/peri/assets/52165881/e2fad58b-05ad-4781-ad4c-e6dcf7dfd5d2)

Then after clicking the `Mark button` the shift back will still be:
![image](https://github.com/IraSoro/peri/assets/52165881/4eb55e72-4815-4ffe-8cab-c1320c7786a0)

IraSoro reviewed (changes requested):
I found a bug.
Steps:
1. Set San Francisco location
2. Set empty array
3. Open `Welcome modal` and mark these days:
![image](https://github.com/IraSoro/peri/assets/52165881/1f73a987-5f82-4c9b-9fd9-43fb5e39ace0)

4. Click `continue` and move to the `tab Home`:
![image](https://github.com/IraSoro/peri/assets/52165881/ba6528df-1df9-4242-9a49-aaa22c6cdcf2)

5. Click `mark button`
![image](https://github.com/IraSoro/peri/assets/52165881/2c019c52-74f5-4c21-9338-c0a32903febc)

There was a shift back 1 day. This only happens in the San Francisco location. 

And the array can be made not empty. You can, for example, uncomment the `_randomLutealPhase(6)` function (or `_randomDelayOfCycle(6) or _todayPeriod(6));` function):
![image](https://github.com/IraSoro/peri/assets/52165881/e2fad58b-05ad-4781-ad4c-e6dcf7dfd5d2)

Then after clicking the `Mark button` the shift back will still be:
![image](https://github.com/IraSoro/peri/assets/52165881/4eb55e72-4815-4ffe-8cab-c1320c7786a0)

IraSoro reviewed (approved):
I didn't find any bugs
